### ISSUE TITLE
Using Literal to over-constrain an Enum does not validate JSON correctly

### ISSUE THREAD

**Issue by willstott101** [pydantic/pydantic#8708]
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

Unfortunately pydantic won't coerce "a_backend" from json into Backends.A for AModel even though it would for CommonModel. If pydantic accepts Literal with enum values at model build time I would expect it to be more aligned with json behaviour for enums elsewhere.

This seems like a bug since in this example the JSON schema looks correct for what I was trying to achieve.

### Example Code

```Python
from pydantic import BaseModel
from typing import Literal
from enum import Enum
from pprint import pprint

class Backends(Enum):
    A = "a_backend"
    B = "b_service"
    C = "smile"

class CommonModel(BaseModel):
    backend: Backends
    common_arg: int

class AModel(CommonModel):
    backend: Literal[Backends.A]
    a_specific_arg: int

class BCModel(CommonModel):
    backend: Literal[Backends.B, Backends.C]
    b_or_c_specific_arg: str

class Request(BaseModel):
    config: AModel | BCModel

pprint(BCModel.model_json_schema())

BCModel.model_validate_json('{"backend": "b_service", "b_or_c_specific_arg": ":)", "common_arg": 12}')
```

### Python, Pydantic & OS Version

```Text
pydantic version: 2.6.0
        pydantic-core version: 2.16.1
          pydantic-core build: profile=release pgo=true
                 install path: /home/wstott/repos/tritium/cloud/service_proxy/venv/lib/python3.10/site-packages/pydantic
               python version: 3.10.12 (main, Nov 20 2023, 15:14:05) [GCC 11.4.0]
                     platform: Linux-6.5.0-14-generic-x86_64-with-glibc2.35
             related packages: typing_extensions-4.9.0 pyright-1.1.349 mypy-1.8.0
                       commit: unknown
```

**Comment by sydney-runkle** [comment]
@willstott101,

This has been discussed previously in https://github.com/pydantic/pydantic/issues/3850 and https://github.com/pydantic/pydantic/issues/2257.

There's a pretty simple solution, which is inheriting from both `str` and `Enum` in your `Backends` class. That being said, this issue has come up enough that I think a docs update would help to make this clear for users in the future.

See this example for how to resolve your problem:

```py
from pydantic import BaseModel
from typing import Literal
from enum import Enum
from pprint import pprint

class Backends(str, Enum): # note, subclassing from str as well
    A = "a_backend"
    B = "b_service"
    C = "smile"

class CommonModel(BaseModel):
    backend: Backends
    common_arg: int

class AModel(CommonModel):
    backend: Literal[Backends.A]
    a_specific_arg: int

class BCModel(CommonModel):
    backend: Literal[Backends.B, Backends.C]
    b_or_c_specific_arg: str

class Request(BaseModel):
    config: AModel | BCModel

pprint(BCModel.model_json_schema())

BCModel.model_validate_json('{"backend": "b_service", "b_or_c_specific_arg": ":)", "common_arg": 12}')
```

**Comment by sydney-runkle** [comment]
I'll also note, our existing docs recommend that you use pure `Literal` types for performance: https://docs.pydantic.dev/latest/concepts/performance/#use-literal-not-enum

**Comment by sydney-runkle** [comment]
I've attached this issue to the common mistakes docs section issue - I think that's where an update would make the most sense. That new docs section should be included in our 2.7 release.
